### PR TITLE
Make the high-level REST client skip types for document deletion.

### PR DESCRIPTION
**Pull request by jpountz**
I'm starting with delete since it is a bit simpler than other APIs, but we
should eventually do the same with all other APIs that are being replaced with
a typeless version (get, put_mapping, search, etc.).

Relates #15613

### CODE REVIEW

**Comment by javanna** [comment]
Is the plan to migrate to new requests/responses as part of this change, or is that part necessary to not also affect transport client? I wonder because that's a breaking change for users, and I wonder what the plan is for all other requests/responses. Also, if we go ahead this way, we may want to remove response parsing code from `DeleteResponse` and add unit tests for the new response class, as well as unit test the request, its validation bits etc.

**Comment by jpountz** [comment]
Thanks for the feedback @javanna, this is the kind of comments that I was looking for. :)

> Is the plan to migrate to new requests/responses as part of this change, or is that part necessary to not also affect transport client?

It would be best if we could migrate the existing request/response classes, but unfortunately this is hard to do because types are only deprecated on the REST layer, not removed yet so the transport client needs to retain a way to set types on request objects, and there is already an `IndexRequest(String,String)` constructor where the second string is the type and the id autogenerated, while we'd want this constructor to take the id as a second parameter, so we can't nicely migrate users from the old API to the new one in a backward-compatible way with nice deprecation warnings. My understanding is that we are generally more keen to break the Java API than the REST layer since the compiler can help figure out what is wrong. Since decoupling the client from the server seems to be something we would like to do in general, creating new request/response classes felt like the right thing to do.

> if we go ahead this way, we may want to remove response parsing code from DeleteResponse and add unit tests for the new response class, as well as unit test the request, its validation bits etc.

+1 We can't remove parsing logic from the old DeleteResponse class yet because of bulk requests, but I agree we should do this eventually. I'll add tests to the request/response classes.

**Comment by javanna** [comment]
> My understanding is that we are generally more keen to break the Java API than the REST layer since the compiler can help figure out what is wrong. 

True, but if breaking was almost required for the transport client given the internal nature of the exposed classes, I hope at some point the REST client becomes stable. Obviously we are not there yet. Let's document this clearly and make sure that users know what to expect, it is true that the compiler helps but it's also good for users to know what they have to do to migrate before they go and do it. In a way it's good to show the direction, the delete API was added with 5.6 or earlier, and it will be updated to its own classes in 7.0 which is not bad at all.

**Comment by jpountz** [comment]
@javanna I pushed one more commit to add tests and document the breaking change.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/DeleteResponseTests.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.index.seqno.SequenceNumbers;
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.IOException;
+import java.util.Collections;
+
+public class DeleteResponseTests extends ESTestCase {
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/DeleteResponseTests.java]: testing these new classes requires to repeat a lot of similar checks, that would be avoided with a test class like AbstractXContentTestCase. But this new class does not implement toXContent hence it's hard to do that. I have faced something similar in #33176 and added a possible approach. The bigger problem though is that we need to make sure that all the response parser support unknown fields (for forward compatibility) and test that, which is something that is very easy to forget if we extend ESTestCase. @hub-cap said he needs to take some time to think about this testing problem caused by using different objects in the client.

**Comment by javanna** [comment]
Thanks for updating! I am good with this but there's a question mark on testing which requires @hub-cap 's thoughts, and his opinion on the overall direction would be great to have too.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/DeleteResponseTests.java (hunk 1 above)
Comment (jpountz) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/DeleteResponseTests.java]: > #33176

Thanks for linking.

> The bigger problem though is that we need to make sure that all the response parser support unknown fields

Indeed I actually explicitly disabled support for unknown fields, which is a mistake now that I read your comment.

**Comment by hub-cap** [comment]
WRT the testing, we will have a conclusion very soon. Stay tuned.

**Comment by hub-cap** [comment]
WRT breaking the existing API, im ok also with this until we get to 7.0, since its not technically released. Ill let others tell me this is a wrong approach tho cc @jasontedor for weighing in.

**Comment by javanna** [comment]
> WRT breaking the existing API, im ok also with this until we get to 7.0, since its not technically released.

@hub-cap just one small question, what do you mean with "its not technically released"?

**Comment by hub-cap** [comment]
@javanna correct me if im wrong, but I assumed that we were "releasing" this in 7.0, and things prior to 7.0 were ok to change. If this is not the case then I definitely need to adjust my way of thinking!!! :)

**Comment by javanna** [comment]
You should try and tell users that @hub-cap  :) I guess for them , what they can use is released?

**Comment by hub-cap** [comment]
haha, yea thats a fair point. I guess usable vs guaranteed back compat are 2 different things. Ill figure out what is the actual answer here.

**Comment by javanna** [comment]
I think we can and have to break things, but let's be aware of when we do so and try to document such changes and help users to migrate? After all they have just made the effort of moving to our new client :)

**Comment by jtibshirani** [comment]
@jpountz would it be okay to close this, as we've decided to take a different approach to types deprecation and the Java HLRC? In particular, we are preferring to add deprecations to existing methods, as opposed to creating new requests + responses for every affected API.

**Comment by jpountz** [comment]
Absolutely.

**Pull request closed without merging**